### PR TITLE
Support IBMBackend.run()

### LINKED ISSUE #1117: Support backend.run() in IBM Cloud channel
We are moving the Researchers program from IQP to IBM Cloud. Since researchers are likely to be interested in functionalities provided by backend.run (e.g. dynamic circuits), we want to support `backend.run()` on the IBM Cloud channel as well.

Since `backend.run` [inputs are changing](https://github.com/Qiskit/qiskit-ibm-provider/issues/640), `backend.run` in `qiskit-ibm-runtime` should just accept new inputs only. People who want to use the old can continue to use `qiskit-ibm-provider`. 

Note that `backend.run()` may become a third primitive, and the interface may change

### PR DESCRIPTION
**Pull request by merav-aharoni**
### Summary
The `backend.run()` functionality was copied from `qiskit-ibm-provider`. I made adjustment as needed. Also copied all the tests that I thought were relevant.
A few open issues:

- [x] Should we use `qiskit-ibm-runtime.Options` for `IBMBackend`?
- [x] Unite `Session` from `provider` with `Session` here. Currently have 2 separate classes - TBD in #1180.
- [x] Do we want to unite `test-ibm-job.py` with `test-job.py`? or keep it separate and change the name? - for now keeping both.
- [x] Do we want to keep all the `status` values from `qiskit-ibm-provider`?
- [x] Unite exceptions from `qiskit-ibm-provider` with those here - do when the provider is merged into runtime.
- [x] Are the tests in `test_ibm_job.py` redundant wrt existing tests here? - for now keeping all tests here.
- [x] `test/fake_account_client.py` do we need the methods here or should they be replaced by something already here? - issue #1160
- [x] Do we want to support `statuses` in `Qiskit.RuntimeService.jobs()` as in `provider.backend.jobs()`? See for example `test_retrieve_jobs_with_status` in the provider. - No, see @jyu00 's comment below.
- [x] `RuntimeJob.test_refresh_job_result`, `test_job_tags_replace` skipped - how do we support `refresh` in runtime? - issue #1159.
- [x] `RuntimeJob.test_queue_info` - Do we want to support `queue_info` in runtime? - issue #1161.
- [x] `RuntimeJob.time_per_step` - do we want to support it? - on hold for now.
- [x] `transpiler/passmanger` directory was copied over to support `_deprecate_id_instruction`, which is already deprecated. Can we remove? - removed `_deprecate_id_instruction`. 
- [x] Tutorial - start by writing documentation - issue #1181.

### Details and comments
Fixes #1117

### CODE REVIEW

**Comment by merav-aharoni** [comment]
@jyu00 , @kt474 - I would appreciate your inputs on the tests for this PR before I continue. My approach has been the following:
I am transferring all the tests that seem relevant from `qiskit-ibm-provider` and making adjustments as needed. So far, I transferred:

- all tests from `provider/test/integration/test_backend.py` to `runtime/test/integration/test_backend.py`
- the complete file `provider/integration/test_ibm_job.py` as a new integration test file
- the complete file `provider/test/unit/test_backend.py` as a new unit test file
- all tests from  `provider/test/unit/test_ibm_job_state.py` to `runtime/test/unit/test_ibm_job_state.py`.
Many of the changes I made in this PR were required to run all these tests.
I am mostly looking at test files that have `backend.run()` and transferring those tests, and others that seem relevant (e.g., related to `job`).

- Any comments on the above?
- Any additional tests you can think of? For example, should we do something similar to `test_ibm_primitives.py`?

**Comment by jyu00** [comment]
A few open issues:
> 
> * [ ]  Should we use `qiskit-ibm-runtime.Options` for `IBMBackend`?

No.
> * [ ]  Do we want to keep all the `status` values from `qiskit-ibm-provider`?

Are you referring to job status? Don't they come from Qiskit? 

> * [ ]  `test/fake_account_client.py` do we need the methods here or should they be replaced by something already here?

Replacing with something already here seems easier? 

> * [ ]  Do we want to support `statuses` in `Qiskit.RuntimeService.jobs()` as in `provider.backend.jobs()`? See for example `test_retrieve_jobs_with_status` in the provider.

No. Qiskit Runtime API doesn't support filtering by granular statuses, so we had to do some complicated stuff in qiskit-ibm-provider for backward compatibility. Since no one has complained about not having this support in qiskit-ibm-runtime, let's just leave out the complexity. 

> * [ ]  `RuntimeJob.test_refresh_job_result`, `test_job_tags_replace` skipped - how do we support `refresh` in runtime?

I guess we can keep `refresh` since it was a [feature request](https://github.com/Qiskit/qiskit-ibmq-provider/issues/466), and it should be fairly easy to do (e.g. just call the API again). 

> * [ ]  `RuntimeJob.test_queue_info` - Do we want to support `queue_info` in runtime?

Yeah, I think people will be more interested in `estimated_start_time` etc once that becomes more accurate lol.

> * [ ]  `RuntimeJob.time_per_step` - do we want to support it?

Let's not do this right now, since it'll change with granular job status.

> * [ ]  `transpiler/passmanger` directory was copied over to support `_deprecate_id_instruction`, which is already deprecated. Can we remove?

I don't understand your comment about the `transpiler/passmanger` directory, but we can remove `_deprecate_id_instruction`.

> Do we want to copy the tests in provider/integration/test_ibm_qasm_simulator.py?

Yes. They are good to ensure if something goes horribly wrong by verifying the results. It'd be good to have similar tests for real backends as well, but result verification could be difficult. 

> Any additional tests you can think of?

How about `test_ibm_job_attributes`? Some of the skipped tests can actually be re-enabled. For example, we can do `test_esp_readout_enabled` since we can retrieve job inputs from runtime API.

**Comment by jyu00** [comment]
> > * [ ]   `RuntimeJob.test_refresh_job_result`, `test_job_tags_replace` skipped - how do we support `refresh` in runtime?
> 
> I guess we can keep `refresh` since it was a [feature request](https://github.com/Qiskit/qiskit-ibmq-provider/issues/466), and it should be fairly easy to do (e.g. just call the API again).

On a second thought, instead of adding a `refresh`, we should just not cache the results and call server again if someone calls `result()` again. Results can be large and caching them in jobs means there is no easy way for a user to delete them to free up memory.

**Comment by merav-aharoni** [comment]
I removed `_deprecate_id_instruction` and the entire `transpiler`directory. I prefer to keep this PR as thin as possible, and to leave open issues to future PRs (I opened several). I think this is a good time for a full review - @kt474 , @jyu00

File: test/integration/test_backend.py [hunk 1]
```diff
@@ -199,3 +199,70 @@ def test_backend_deepcopy(self):
...
+            for hgp in hgps:
+                backend_names = list(hgp._backends)
+                for backend in backends:
+                    if backend.name not in backend_names:
+                        with self.assertRaises(QiskitBackendNotFoundError):
+                            self.service.backend(
+                                backend.name,
+                                instance=f"{hgp._hub}/{hgp._group}/{hgp._project}",
+                            )
+                        return
```
Comment (kt474) [inline test/integration/test_backend.py:251]: All of the tests like these that use `._hgps` will have to be updated to be quantum channel only because they will not work on cloud

File: test/integration/test_ibm_job_attributes.py [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+    @classmethod
+    @integration_test_setup()
+    def setUpClass(cls, dependencies: IntegrationTestDependencies) -> None:
+        """Initial class level setup."""
+        # pylint: disable=arguments-differ
+        super().setUpClass()
+        cls.dependencies = dependencies
+        cls.service = dependencies.service
+        cls.sim_backend = dependencies.service.backend(
+            "ibmq_qasm_simulator", instance=dependencies.instance
```
Comment (kt474) [inline test/integration/test_ibm_job_attributes.py:64]: This will also have to updated because the instance keyword is only for the quantum channel (they'll fail anyway right now because `circuit-runner` has not been enabled on the cloud side yet).

File: test/integration/test_ibm_job.py [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+    last_month: datetime
+
+    @classmethod
+    @integration_test_setup_with_backend(simulator=False, min_num_qubits=2)
+    def setUpClass(cls, backend: IBMBackend, dependencies: IntegrationTestDependencies) -> None:
+        """Initial class level setup."""
+        # pylint: disable=arguments-differ
+        super().setUpClass(dependencies=dependencies)
+        cls.sim_backend = dependencies.service.backend(
+            "ibmq_qasm_simulator", instance=dependencies.instance
```
Comment (kt474) [inline test/integration/test_ibm_job.py:62]: Same thing here, we should remove the instance parameter here bc it won't work with the cloud channel

File: test/integration/test_ibm_qasm_simulator.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+class TestIBMQasmSimulator(IBMIntegrationTestCase):
+    """Test IBM Quantum QASM Simulator."""
+
+    @integration_test_setup_with_backend(simulator=False)
+    def setUp(self, backend: IBMBackend, dependencies: IntegrationTestDependencies) -> None:
+        """Initial test setup."""
+        # pylint: disable=arguments-differ
+        super().setUp()
+        self.sim_backend = self.service.backend(
+            "ibmq_qasm_simulator", instance=dependencies.instance
```
Comment (kt474) [inline test/integration/test_ibm_qasm_simulator.py:42]: Same thing here

File: qiskit_ibm_runtime/ibm_backend.py [hunk 5]
```diff
@@ -549,6 +581,272 @@ def __deepcopy__(self, _memo: dict = None) -> "IBMBackend":
...
+        shots: Optional[Union[int, float]] = None,
+        memory: Optional[bool] = None,
+        meas_level: Optional[Union[int, MeasLevel]] = None,
+        meas_return: Optional[Union[str, MeasReturnType]] = None,
+        rep_delay: Optional[float] = None,
+        init_qubits: Optional[bool] = None,
+        use_measure_esp: Optional[bool] = None,
+        noise_model: Optional[Any] = None,
+        seed_simulator: Optional[int] = None,
+        **run_config: Dict,
```
Comment (kt474) [inline qiskit_ibm_runtime/ibm_backend.py:601]: Do we want to remove `memory`, and `meas_return` as well as move `noise_model` and `seed_simulator` to the backend options as Blake mentioned [here](https://github.com/Qiskit/qiskit-ibm-provider/issues/640)?

And then have `memory=True` and `meas_return=MeasReturnType.SINGLE` as the default 
@jyu00

File: test/integration/test_ibm_job_attributes.py (hunk 2 above)
Comment (merav-aharoni) [reply test/integration/test_ibm_job_attributes.py]: Done.

File: test/integration/test_ibm_job.py (hunk 3 above)
Comment (merav-aharoni) [reply test/integration/test_ibm_job.py]: Done.

File: test/integration/test_ibm_qasm_simulator.py (hunk 4 above)
Comment (merav-aharoni) [reply test/integration/test_ibm_qasm_simulator.py]: Done.

**Comment by merav-aharoni** [comment]
@kt474 - how can I run the integration tests in CI?

**Comment by kt474** [comment]
> @kt474 - how can I run the integration tests in CI?

I made some minor updates, now all tests are passing in both cloud and quantum channels 

I usually just test locally with `make integration-test`. If you want to test in the CI you could remove the line [here](https://github.com/Qiskit/qiskit-ibm-runtime/blob/main/.github/workflows/integration-tests.yml#L21) and then run the tests from your fork (for example my [fork](https://github.com/kt474/qiskit-ibm-runtime/actions/workflows/integration-tests.yml))

File: qiskit_ibm_runtime/ibm_backend.py [hunk 6]
```diff
@@ -43,10 +46,20 @@
...
 from qiskit_ibm_provider.utils import local_to_utc
+from qiskit_ibm_provider.utils import validate_job_tags, are_circuits_dynamic
+from qiskit_ibm_provider.utils.options import QASM2Options, QASM3Options
+from qiskit_ibm_provider.exceptions import IBMBackendValueError, IBMBackendApiError
+from qiskit_ibm_provider.api.exceptions import RequestsApiError
+
+# temporary until we unite the 2 Session classes
+from qiskit_ibm_provider.session import (
+    Session as ProviderSession,
+)  # temporary until we unite the 2 Session classes
```
Comment (kt474) [inline qiskit_ibm_runtime/ibm_backend.py:57]: Yeah I think this is fine for now since we are not removing any functionality 

```python
backend = service.backend('ibmq_qasm_simulator')

with Session(service, backend=backend) as session:
    sampler = Sampler(session=session)
    
    job1 = sampler.run(bell)
    job2 = backend.run(bell) # this would not run in a session 

```
We can still use `backend.run()` in sessions like before but we should unite/consildate the classes so this code snippet works.

File: qiskit_ibm_runtime/constants.py [hunk 7]
```diff
@@ -23,10 +23,21 @@
...
+    "VALIDATED": JobStatus.VALIDATING,
     "QUEUED": JobStatus.QUEUED,
+    "PENDING_IN_QUEUE": JobStatus.QUEUED,
     "RUNNING": JobStatus.RUNNING,
     "COMPLETED": JobStatus.DONE,
     "FAILED": JobStatus.ERROR,
+    "ERROR_CREATING_JOB": JobStatus.ERROR,
+    "ERROR_VALIDATING_JOB": JobStatus.ERROR,
+    "ERROR_RUNNING_JOB": JobStatus.ERROR,
+    "ERROR_TRANSPILING_JOB": JobStatus.ERROR,
```
Comment (jyu00) [inline qiskit_ibm_runtime/constants.py:40]: We don't get these additional statuses anymore. They were from the old Qobj path.

File: qiskit_ibm_runtime/ibm_backend.py (hunk 5 above)
Comment (jyu00) [reply qiskit_ibm_runtime/ibm_backend.py:601]: Let's keep things the same for now. We can make the changes when we overhaul `backend.run`.

File: qiskit_ibm_runtime/ibm_backend.py [hunk 8]
```diff
@@ -492,10 +509,25 @@ def __call__(self) -> "IBMBackend":
...
+            IBMBackendValueError:
+                - If one of the circuits contains more qubits than on the backend."""
+
+        if len(circuits) > self._max_circuits:
+            raise IBMBackendValueError(
+                f"Number of circuits, {len(circuits)} exceeds the "
+                f"maximum for this backend, {self._max_circuits})"
+            )
+        for circ in circuits:
+            if isinstance(circ, QuantumCircuit):
```
Comment (jyu00) [inline qiskit_ibm_runtime/ibm_backend.py:524]: Type hint needs to be fixed if the input isn't always `QuantumCircuit`

File: qiskit_ibm_runtime/ibm_backend.py [hunk 9]
```diff
@@ -549,6 +581,274 @@ def __deepcopy__(self, _memo: dict = None) -> "IBMBackend":
...
+                if key not in fields and not self.configuration().simulator:
+                    warnings.warn(  # type: ignore[unreachable]
+                        f"{key} is not a recognized runtime option and may be ignored by the backend.",
+                        stacklevel=4,
+                    )
+            elif backend_options.get(key) is not None and key in fields:
+                run_config_dict[key] = backend_options[key]
+        return run_config_dict
+
+    def open_session(self, max_time: Optional[Union[int, str]] = None) -> ProviderSession:
```
Comment (jyu00) [inline qiskit_ibm_runtime/ibm_backend.py:833]: we have an `open_session` but no `close_session` ?

File: qiskit_ibm_runtime/qiskit_runtime_service.py [hunk 10]
```diff
@@ -1344,6 +1346,8 @@ def jobs(
                     "The 'instance' keyword is only supported for ``ibm_quantum`` runtime."
                 )
             hub, group, project = from_instance_format(instance)
+        if job_tags:
+            validate_job_tags(job_tags, IBMBackendValueError)
```
Comment (jyu00) [inline qiskit_ibm_runtime/qiskit_runtime_service.py:1350]:
```suggestion
            validate_job_tags(job_tags, ValueError)
```

In fact we can probably just hard code `ValueError` in `validate_job_tags`. I don't remember why it was done this way originally.

**Review by jyu00 — COMMENTED** [review]
We'll need to add the transpiler passes (see [here](https://github.com/Qiskit/qiskit-ibm-provider/blob/main/setup.py#L96-L102))

File: qiskit_ibm_runtime/ibm_backend.py (hunk 6 above)
Comment (merav-aharoni) [reply qiskit_ibm_runtime/ibm_backend.py:57]: If you agree, I'd rather leave this for a separate PR.

File: qiskit_ibm_runtime/ibm_backend.py (hunk 8 above)
Comment (merav-aharoni) [reply qiskit_ibm_runtime/ibm_backend.py:524]: Actually, this check is unnecessary, so I removed the `isinstance` line.

Comment (merav-aharoni) [reply qiskit_ibm_runtime/ibm_backend.py:522]: On further checking, there is one test `test_single_openqasm3_submission` that submits a string, so I put back the check and updated the hint.

File: qiskit_ibm_runtime/ibm_backend.py (hunk 9 above)
Comment (merav-aharoni) [reply qiskit_ibm_runtime/ibm_backend.py:831]: Added `IBM_Backend.close_session` in addition to `IBM_Backend.cancel_session` in https://github.com/Qiskit/qiskit-ibm-provider/pull/762. This needs to be merged before we can use it here.

File: qiskit_ibm_runtime/qiskit_runtime_service.py (hunk 10 above)
Comment (merav-aharoni) [reply qiskit_ibm_runtime/qiskit_runtime_service.py]: Removed the `exception` parameter. Use `validate_job_tags` from this repo instead of from the provider.

File: setup.py [hunk 11]
```diff
@@ -77,4 +77,10 @@
         "Documentation": "https://qiskit.org/documentation/",
         "Source Code": "https://github.com/Qiskit/qiskit-ibm-runtime",
     },
+    entry_points={
+        "qiskit.transpiler.translation": [
+            "ibm_backend = qiskit_ibm_provider.transpiler.plugin:IBMTranslationPlugin",
+            "ibm_dynamic_circuits = qiskit_ibm_provider.transpiler.plugin:IBMDynamicTranslationPlugin",
```
Comment (jyu00) [inline setup.py:83]:
```suggestion
            "ibm_backend = qiskit_ibm_provider.transpiler.plugin:IBMTranslationPlugin",
            "ibm_dynamic_circuits = qiskit_ibm_provider.transpiler.plugin:IBMDynamicTranslationPlugin",
```
```suggestion
            "ibm_backend = qiskit_ibm_runtime.transpiler.plugin:IBMTranslationPlugin",
            "ibm_dynamic_circuits = qiskit_ibm_runtime.transpiler.plugin:IBMDynamicTranslationPlugin",
```

please double check the paths are correct

File: qiskit_ibm_runtime/ibm_backend.py [hunk 12]
```diff
@@ -492,10 +507,25 @@ def __call__(self) -> "IBMBackend":
         # For backward compatibility only, can be removed later.
         return self
 
-    def run(self, *args: Any, **kwargs: Any) -> None:
-        """Not supported method"""
-        # pylint: disable=arguments-differ
-        raise RuntimeError("IBMBackend.run() is not supported in the Qiskit Runtime environment.")
+    def _check_circuits_attributes(self, circuits: Union[List[QuantumCircuit], str]) -> None:
```
[the remark this answers is not in the document]
Comment (merav-aharoni) [reply qiskit_ibm_runtime/ibm_backend.py]: I agree.

File: qiskit_ibm_runtime/ibm_backend.py [hunk 13]
```diff
@@ -549,6 +579,274 @@ def __deepcopy__(self, _memo: dict = None) -> "IBMBackend":
...
+        hgp_name = None
+        if self._service._channel == "ibm_quantum":
+            hgp_name = self._instance or self._service._get_hgp().name
+
+        session = self._session
+
+        if session:
+            if not session.active:
+                raise RuntimeError(f"The session {session.session_id} is closed.")
+            session_id = session.session_id or None
```
[the remark this answers is not in the document]
Comment (merav-aharoni) [reply qiskit_ibm_runtime/ibm_backend.py:762]: You are right. Changed this section of the code.

File: qiskit_ibm_runtime/ibm_backend.py [hunk 14]
```diff
@@ -549,6 +579,274 @@ def __deepcopy__(self, _memo: dict = None) -> "IBMBackend":
...
+            max_execution_time = session._max_time
```
[the remark this answers is not in the document]
Comment (merav-aharoni) [reply qiskit_ibm_runtime/ibm_backend.py:763]: Removed `max_execution_time` and changed this section of code.

File: setup.py (hunk 11 above)
Comment (merav-aharoni) [reply setup.py]: Checked and replaced.

File: qiskit_ibm_runtime/ibm_backend.py [hunk 15]
```diff
@@ -549,6 +579,274 @@ def __deepcopy__(self, _memo: dict = None) -> "IBMBackend":
...
+            use_measure_esp=use_measure_esp,
+            noise_model=noise_model,
+            seed_simulator=seed_simulator,
+            **run_config,
+        )
+
+        run_config_dict["circuits"] = circuits
+        if not program_id.startswith(QASM3RUNNERPROGRAMID):
+            # Transpiling in circuit-runner is deprecated.
+            run_config_dict["skip_transpilation"] = True
```
[the remark this answers is not in the document]
Comment (merav-aharoni) [reply qiskit_ibm_runtime/ibm_backend.py]: Removed.

File: qiskit_ibm_runtime/ibm_backend.py [hunk 16]
```diff
@@ -549,6 +579,274 @@ def __deepcopy__(self, _memo: dict = None) -> "IBMBackend":
...
+                log_level=log_level,
+                job_tags=job_tags,
+                session_id=session_id,
+                start_session=start_session,
+                max_execution_time=max_execution_time,
+                image=image,
+            )
+        except RequestsApiError as ex:
+            raise IBMBackendApiError("Error submitting job: {}".format(str(ex))) from ex
+        session_id = response.get("session_id")
```
[the remark this answers is not in the document]
Comment (merav-aharoni) [reply qiskit_ibm_runtime/ibm_backend.py:775]: I wrote this similar to the line in [runtime_job.py](https://github.com/Qiskit/qiskit-ibm-runtime/blob/ccd41a529f065fd0c50087a8238f44c986bfbe7c/qiskit_ibm_runtime/runtime_job.py#L650).
Also, I tried printing the `session_id` before and after this line and indeed the session_id is returned.
To make sure, I added `None` as follows:
`session_id = response.get("session_id", None)`

File: qiskit_ibm_runtime/constants.py (hunk 7 above)
Comment (merav-aharoni) [reply qiskit_ibm_runtime/constants.py]: Okay, removed the extra values.

**Comment by merav-aharoni** [comment]
@jyu00 - I addressed all your comments from yesterday. Please have a look.

File: qiskit_ibm_runtime/ibm_backend.py [hunk 17]
```diff
@@ -549,6 +579,263 @@ def __deepcopy__(self, _memo: dict = None) -> "IBMBackend":
...
+
+        image: Optional[str] = run_config.get("image", None)  # type: ignore
+        if image is not None:
+            image = str(image)
+
+        if isinstance(init_circuit, bool):
+            warnings.warn(
+                "init_circuit does not accept boolean values. "
+                "A quantum circuit should be passed in instead."
+            )
```
Comment (jyu00) [inline qiskit_ibm_runtime/ibm_backend.py:709]: (can be separate PR) - this should be an error and not a warning.

File: qiskit_ibm_runtime/ibm_backend.py [hunk 18]
```diff
@@ -549,6 +579,263 @@ def __deepcopy__(self, _memo: dict = None) -> "IBMBackend":
...
+        """Runs the runtime program and returns the corresponding job object"""
+        hgp_name = None
+        if self._service._channel == "ibm_quantum":
+            hgp_name = self._instance or self._service._get_hgp().name
+
+        session = self._session
+        if session and not session.active:
+            raise RuntimeError(f"The session {session.session_id} is closed.")
+        session_id = session.session_id if session else None
+        start_session = session_id is None
```
Comment (jyu00) [inline qiskit_ibm_runtime/ibm_backend.py:758]: Wouldn't this also be `True` if I never passed in a Session ?

File: qiskit_ibm_runtime/ibm_backend.py [hunk 19]
```diff
@@ -549,6 +579,263 @@ def __deepcopy__(self, _memo: dict = None) -> "IBMBackend":
...
+                        f"{key} is not a recognized runtime option and may be ignored by the backend.",
+                        stacklevel=4,
+                    )
+            elif backend_options.get(key) is not None and key in fields:
+                run_config_dict[key] = backend_options[key]
+        return run_config_dict
+
+    def open_session(self, max_time: Optional[Union[int, str]] = None) -> ProviderSession:
+        """Open session"""
+        self._session = ProviderSession(max_time)
```
Comment (jyu00) [inline qiskit_ibm_runtime/ibm_backend.py:822]: This can also go to a separate PR, but I don't see where `max_time` is passed to the server.

File: qiskit_ibm_runtime/ibm_backend.py (hunk 17 above)
Comment (merav-aharoni) [reply qiskit_ibm_runtime/ibm_backend.py]: Changed to an error. I didn't think it was worth adding a test.

File: qiskit_ibm_runtime/ibm_backend.py (hunk 18 above)
Comment (merav-aharoni) [reply qiskit_ibm_runtime/ibm_backend.py]: You are right. Added a check for whether there is a session.

File: qiskit_ibm_runtime/ibm_backend.py (hunk 19 above)
Comment (merav-aharoni) [reply qiskit_ibm_runtime/ibm_backend.py]: Correct. I removed `max_time` and will opened a new PR to decide how to handle this.

**Review by kt474 — APPROVED** [review]

File: qiskit_ibm_runtime/ibm_backend.py [hunk 20]
```diff
@@ -755,7 +755,7 @@ def _runtime_run(
         if session and not session.active:
             raise RuntimeError(f"The session {session.session_id} is closed.")
         session_id = session.session_id if session else None
-        start_session = session_id is None
+        start_session = session is not None and session_id is None
```
Comment (jyu00) [inline qiskit_ibm_runtime/ibm_backend.py:758]: While the code works, I think it'd be easier to follow if we just group all the session related testing inside an `if` state, e.g. 

```
if self._session:
    if not session.active:
        raise ..
    session_id = self._session.session_id
    start_session = session_id is None
else:
    session_id = None
    start_session = False
```

**Review by jyu00 — APPROVED** [review]
One small comment but otherwise LGTM

**Comment by kt474** [comment]
Thanks @merav-aharoni! 🎉
